test/membrane/error.spec.js

```diff
@@ -0,0 +1,212 @@
+
+                try {
+                    foo()
+                } catch(e) {
+                    expect(e.foo).toBe('bar')
+                    expect(e).toBe(errorObj)
+                }
+            `)
+})
+
+it('non-error objects thrown in red constructors', () => {
+    const evalScript = createSecureEnvironment(undefined, { expect })
+    evalScript(`
+                const errorObj = { foo: 'bar' }
+                
+                class Foo {
+                    constructor() {
+                        throw errorObj;
+                    }
+                }
+
+                try {
+                    new Foo()
+                } catch(e) {
+                    expect(e.foo).toBe('bar')
+                    expect(e).toBe(errorObj)
+                }
+            `)
+})
+
+it('non-error objects thrown in Promise', (done) => {
+    const evalScript = createSecureEnvironment(undefined, { done, expect })
+    evalScript(`
+                const error = { foo: 'bar' }
+                const p = new Promise(() => {
+                    throw error
+                })
+                p.catch((e) => {
+                    expect(e.foo).toBe('bar')
+                    expect(e).toBe(error)
+                    done()
+                })
+            `)
+})
+
+it('unhandled promise rejections with non-error objects', (done) => {
+    const evalScript = createSecureEnvironment(undefined, { done, expect })
+    evalScript(`
+        const errorObj = { foo: 'bar' }
+        window.addEventListener("unhandledrejection", event => {
+            expect(event.reason).toBe(errorObj)
+            expect(event.reason.foo).toBe('bar')
+            done()
+        });
+
+        new Promise((resolve, reject) => {
+            throw errorObj
+        });
+    `)
+})
+
+it('non-error objects thrown in blue functions do not leak', () => {
+    function foo() {
+        throw { foo: 'bar' }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { foo, expect })
+    evalScript(`
+                try {
+                    foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in blue constructors do not leak', () => {
+    class Foo {
+        constructor() {
+            throw { foo: 'bar' }
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+                try {
+                    new Foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in red functions do not leak', () => {
+    let fn
+    function save(arg) {
+        fn = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`
```

Comment:
Don't get why pass save?

Reply:
seems like a faulty test... but it does make sense why it passes... jajajaja... because `fn` is undefined (never saved). maybe we should make sure that the error has the proper message.

Reply:
sorry, missed the call to `save` function. Added now.

Reply:
Also validating the message is not defined (eliminates all other errors)